### PR TITLE
CIS194-Week4-Exercise

### PR DESCRIPTION
**Pull request by nielstj**

### CODE REVIEW

File: week4/exercise.hs [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+module Week4 where
+
+-- Exercise 1 : Wholemeal programming
+
+fun1' :: [Integer] -> Integer
+fun1' =  product . map (subtract 2) . filter even
+
+fun2' :: Integer -> Integer
+fun2' = sum . filter (even) . takeWhile (/=1) . iterate next
```
Comment (chshersh) [inline week4/exercise.hs]: `even` is not an operator, so you don't need `()` here.
```suggestion
fun2' = sum . filter even . takeWhile (/=1) . iterate next
```

File: week4/exercise.hs [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+  where
+    next :: Integer -> Integer
+    next x = if even x then div x 2 else 3 * x + 1
```
Comment (chshersh) [inline week4/exercise.hs:14]: Some people use infix version of the `div` function, like this:

```haskell
x `div` 2
```

But that depends on personal taste 🙂

File: week4/exercise.hs [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+        height Leaf = -1
+        height (Node h _ _ _) = h
+
+foldTree :: [a] -> Tree a
+foldTree = foldr insert' Leaf 
+
+-- Exercise 3 : More folds
+
+xor :: [Bool] -> Bool
+xor = foldr1 (/=)
```
Comment (chshersh) [inline week4/exercise.hs]: `foldr1` is a partial function, so this particular implementation will fail on empty lists. Instead, it's better to use`foldr`. You only need to think of a default to make it work with the `/=` operator.

**Review by chshersh — COMMENTED** [review]
Nice implementation! I have only minor comments. But I see a very good understanding of Haskell concepts 👏

**Comment by nielstj** [comment]
> Nice implementation! I have only minor comments. But I see a very good understanding of Haskell concepts 👏

Thank you!

**Pull request merged**